src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncodeInternal(string originalString)
+        {
+            fixed (char* pch = originalString)
+            {
+                int safeCharCount = 0;
+                int spaceCharCount = 0;
+                for (int i = 0; i < originalString.Length; i++)
+                {
+                    char c = pch[i];
+                    if (c == ' ')
+                    {
+                        spaceCharCount++;
+                    }
+                    else if (IsUrlSafeChar(c))
+                    {
+                        safeCharCount++;
+                    }
+                }
+
+                if (spaceCharCount == 0 && safeCharCount == originalString.Length)
+                {
+                    // Nothing to expand
+                    return originalString;
+                }
+                
+                int unsafeByteCount = Encoding.UTF8.GetByteCount(pch, originalString.Length) - safeCharCount - spaceCharCount;
```

Comment:
@hughbe Thanks for replying to my review! What I actually had in mind when I said to use `GetByteCount` was per-character actually, something like this:

```cs
bool foundSpaces = false;
int unsafeCount = 0;

for (int i = 0; i < originalString.Length; i++)
{
    char c = pch[i];
    if (c == ' ') foundSpaces = true;
    else if (!IsUrlSafeChar(c))
    {
        if (i + 1 < originalString.Length && char.IsSurrogatePair(c, pch[i + 1]))
        {
            i++;
            unsafeCount += 4; // Surrogate pairs -> 4 chars
        }
        else
        {
            unsafeCount += Encoding.UTF8.GetByteCount(&c, 1);
        }
    }
}
```

The new implementation you have looks like it makes two passes over the string, one in the `for` loop and one in `GetByteCount`. (See [this](https://github.com/dotnet/corefx/pull/6698/files#diff-831332bbb389db07350457a4d92095f8R325) PR for what I did to handle this.)